src/Dapr.Client/DaprClientGrpc.cs

```diff
@@ -450,6 +451,90 @@ public override async ValueTask<(TValue value, string etag)> GetStateAndETagAsyn
                 cancellationToken);
         }
 
+
+        /// <inheritdoc/>
+        public override async Task ExecuteStateTransactionAsync(
+            string storeName,
+            IReadOnlyList<StateTransactionRequest> operations,
+            Dictionary<string, string> metadata = default,
+            CancellationToken cancellationToken = default)
+        {
+            ArgumentVerifier.ThrowIfNullOrEmpty(storeName, nameof(storeName));
+            ArgumentVerifier.ThrowIfNull(operations, nameof(operations));
+            if (operations.Count == 0)
+            {
+                throw new ArgumentException($"{nameof(operations)} does not contain any elements");
+            }
+
+            await this.MakeExecuteStateTransactionCallAsync(
+                storeName,
+                operations,
+                metadata,
+                cancellationToken);
+        }
+
+        private async ValueTask MakeExecuteStateTransactionCallAsync(
+            string storeName,
+            IReadOnlyList<StateTransactionRequest> states,
+            Dictionary<string, string> metadata = default,
+            CancellationToken cancellationToken = default)
+        {
+            var executeStateTransactionRequestEnvelope = new Autogenerated.ExecuteStateTransactionRequest()
+            {
+                StoreName = storeName,
+            };
+
+            foreach (var state in states)
+            {
+                var stateOperation = new Autogenerated.TransactionalStateOperation();
+
+                stateOperation.OperationType = state.OperationType.ToString().ToLower();
+                stateOperation.Request = ToAutogeneratedStateItem(state);
+
+                executeStateTransactionRequestEnvelope.Operations.Add(stateOperation);
+
+            }
+
+            // Add metadata that applies to all operations if specified
+            if (metadata != null)
+            {
+                executeStateTransactionRequestEnvelope.Metadata.Add(metadata);
+            }
+
+            await this.MakeGrpcCallHandleError(
+                options => client.ExecuteStateTransactionAsync(executeStateTransactionRequestEnvelope, options),
+                cancellationToken);
+        }
+
+        private Autogenerated.StateItem ToAutogeneratedStateItem(StateTransactionRequest state)
+        {
+            var stateOperation = new Autogenerated.StateItem();
+            stateOperation.Key = state.Key;
+
+            if (state.Value != null)
+            {
+                stateOperation.Value = ByteString.CopyFrom(state.Value);
+            }
+
+            if (state.ETag != null)
+            {
+                stateOperation.Etag = state.ETag;
+            }
```

Comment:
Question for both @vinayada1 and @youngbupark - what are the semantics of including an etag here? Does the transaction ignore it? Does the transaction rollback? 

We designed different APIs for the main state store case with ETags - because the application developer *generally needs to know the difference*.

It might be best to make this look more like `TryDeleteStateAsync` https://github.com/dapr/dotnet-sdk/pull/385/files#diff-ca47ccc1f9e29dbb552689c374bf9023L259

Reply:
Now, we didn't specify the exact behavior of etag for transaction.  Rather, the behavior of etag depends on state store implementation.

e.g. Redis state store ignores it for transaction operation. postgre and sql state store rollback the transaction because update is failed. some of state store, aws dynamodb state store doesn't support etag even for single operation and transactional operation at all.

cc/ @yaron2